### ISSUE TITLE
RFE: Track service config files

### ISSUE THREAD

**Issue by purpleidea** [systemd/systemd#4239]
### Submission type
- [ ] Bug report
- [X] Request for enhancement (RFE)

This is an RFE to extend systemd and the unit file (slightly) to improve the experience for the config management and automation communities.
### Proposal

1.) Add a field in the service unit file to declaratively list the config files which the service depends on. This could look something like the following:

```
[Config]
ConfigFile=/etc/foo.conf
ConfigDir=/etc/foo.d/
ConfigMatch=/etc/foo/*.conf
```

You can specify as many of these directives as you'd like. I recommend being able to glob to match simple patterns. These could also all be mixed together under a single directive (eg: `ConfigPath`) if that is decided to be more aesthetically pleasing.

Lennart had mentioned that he would like these to be relative paths, and the absolute prefix (eg: /etc/) be passed in by systemd. This variant doesn't negatively impact this proposal in any way and is up to the systemd core team on which they prefer.

2.) Provide an API to specifically get the last reloaded time of the service in particular. Currently it's possible to get `ExecMainStartTimestamp`, but not `ExecMainReloadTimestamp`. This would be an easy addition to the API.
### Advantage to user

By combining these two concepts, the `systemctl status` command could tell a user if the last modified time of any of the listed config files is newer than the last modified time of reload and start. If so it's likely that the service needs a reload to be able to use the newest version of the configs and be running in the correct state as seen on disk.
### Advantage to config management communities

1.) Currently there's no guaranteed way to detect if a currently running service is in the correct state with respect to the configuration files specified. In particular, if the automation technology updates a configuration file, it will typically then run a reload of the service. However if an error occurs after the file update but before the reload, on a subsequent run, you will loose the information that the service needs reloading!

2.) With the config file information listed in the service unit file, it will be easy for config management tools to automatically know the dependency relationship between the config file and the service. This means they can know to update the config file before starting or reloading the service. This would make everyone's life easier.
### Getting these additions into distribution unit files.

Since this would benefit the sysadmin and devops consumers greatly, and since these types of patches are quite easy to write, even non-programmers using these services will be able to send distribution and upstream patches to add these options. If they are not added, it doesn't hurt the status-quo.
### More information

I'll be presenting this information during my talk at [systemd conf](https://cfp.systemd.io/en/systemdconf_2016/public/events/1), during my [`mgmt`](https://github.com/purpleidea/mgmt/) presentation. I will post the video link once it is available.

Thank you for your consideration!

**Comment by johannbg** [comment]
Afaikt with the former part (1) what you are trying to do there is that you are just pushing something that administrators will need and should express in an configuration management system into type unit files which will just result in increased complexity for type units with the associated overhead for administrators and does not scale. 

The latter (2) is something that could be looked and implemented.

**Comment by purpleidea** [comment]
On Thu, Sep 29, 2016 at 7:40 AM, Jóhann B. Guðmundsson
notifications@github.com wrote:

> Afaikt with the former part (1) what you are trying to do there is that you are just pushing something that administrators will need and should express in an configuration management system into type unit files which will just result in increased complexity for type units with the associated overhead for administrators and does not scale.

Indeed it is something that system administrators need, but currently
it is something they need to dig for, and putting it in the system
init fail makes it the defacto documentation.

Additionally, if systemd doesn't know about this, then it can't tell
the sysadmin when their service needs being restarted. This is
analogous to how systemd notifies you that you might need to
`daemon-reload`.

> The latter (2) is something that could be looked and implemented.

Awesome!

**Comment by purpleidea** [comment]
@poettering Opened as requested. Thanks.

**Comment by johannbg** [comment]
@purpleidea well I would say systemd is not that one that should determine or tell anyone or anything it should be restarted let alone if an file changes in the server(s) but it should tell previous and current state of services etc . 

The complex logic configuration management and associated triggers should reside entirely in the centralized configuration management system ( as well as monitoring systems etc alot of those systemd share similarity in what they do )

If you are managing machines in the thousands and containers in the hundreds of thousands you want a) have centralized configuration management in which the administrators logs in, configures the initial state of the machine(s) as in tells it how it should be partitioned, networked, which components is installed and what services should be running. ( create an template ) 

b) you want to manage the configuration of all those machine in that configuration management system and be able to associate these configuration profiles it to any given template and services within that template ( so you don't end up having to configure it per machine/containers and services on that machine/container as you are proposing in 1 ) 

c) you need to be able to keep track on the changes creation of the configuration(s) and template associated with each machine/container or cluster of containers as in who created/changed, what ( template/configuration ) when and why and what version these templates and configurations are in. 

d)  you want to be able to push and rollback configurations on the machine or container or cluster of machines or containers ( compare the checksum of the configuration file in the configuration management system with what has been deployed in the machine the, container or cluster of machines or containers). 

If the configuration ( or even web applications or databases etc ) has been successfully deployed on the machine, the container or the cluster of machines or containers then start/reload/restart the service(s) ( or rollback encase the checksum does not match on those file(s) and do nothing as not start/reload/restart the service(s) ) or do so as a schedule task as in you might push all configuration changes to the machine(s)/container(s) during normal working hours but want to schedule start/restart/reload the services after hours or when ever the service you provide can be least disruptive to your customer base ( and that might mean one cluster in Germany and another cluster in Australia two different timezone which means two different schedule task to start/reload/restart the service(s)  ) all of which would be managed in the configuration management.

The above is roughly the scope of any modern configuration management system and as such you can view what systemd is lacking in that context as in what dbus interface or api are missing or should be extended to be to support configuration management system ( your number 2 ) but not try to act as one ( as your request 1 is doing which would undoubtedly be followed by some requesting the often requested of turning systemd into syntax checker for configuration files for poorly written daemon/services so it could prevent restarting services with broken configuration  )

**Comment by purpleidea** [comment]
On Fri, Sep 30, 2016 at 11:06 AM, Jóhann B. Guðmundsson
notifications@github.com wrote:

> @purpleidea well I would say systemd is not that one that should determine or tell anyone or anything it should be restarted let alone if an file changes in the server(s) but it should tell previous and current state of services etc .
> 
> The complex logic configuration management and associated triggers should reside entirely in the centralized configuration management system ( as well as monitoring systems etc alot of those systemd share similarity in what they do )

As I explained, it's not possible to solve this problem without the
help of the init system. You can watch a video of my talk that
explains this better here:

https://www.youtube.com/watch?v=jB992Zb3nH0&a&html5=1

Thanks,
James

**Comment by johannbg** [comment]
Was expecting the same boring cms stuff one has seen throughout the years, just another variant written in python ;) but this is actually pretty cool stuff and so cool that it warrants it's own type cms unit so it can be extended in the future and wont clutter and complex the type service unit files any further I would say.
( you might need to order this after specific mount points or order it after condition like ConditionPathIsReadWrite=/etc, context which aren't applicable to the type service file )   

Unlike what you propose ( putting this into the type service unit ) it would just be hooked up to the type service file the usual means as in either by bearing the same name or by expressing it the configuration section

So instead of adding this to foo.service like this

[Config]
ConfigFile=/etc/foo.conf
ConfigDir=/etc/foo.d/
ConfigMatch=/etc/foo/*.conf

You would create the foo.cms and have something like this.
[Unit]
Description=Foo CMS
Documentation=man:foo(1) man:cms(2)
Bla..

[Config]
ConfigFile=/etc/foo.conf
ConfigDir=/etc/foo.d/
ConfigMatch=/etc/foo/*.conf
....

or create bar.cms and have something like this...

You would create the foo.cms and have something like this.
[Unit]
Description=Foo CMS
Documentation=man:foo(1) man:cms(2)
ConditionPathIsReadWrite=/etc

[Config]
Service=foo.service
ConfigFile=/etc/foo.conf
ConfigDir=/etc/foo.d/
ConfigMatch=/etc/foo/*.conf
....

Unless ofcourse there is something that explicitly states this has to be in the type service file?

**Comment by arvidjaar** [comment]
> Add a field in the service unit file to declaratively list the config files which the service depends on.

What ensures that started process is actually using these files?

**Comment by purpleidea** [comment]
On Sat, Oct 1, 2016 at 3:41 AM, Andrey Borzenkov
notifications@github.com wrote:

> Add a field in the service unit file to declaratively list the config files which the service depends on.
> 
> What ensures that started process is actually using these files?

Great question!

The unit service writer who has knowledge of their service knows this
best. If the file is not present on disk, then we obviously don't use
it to compare timestamps to the start/reload times. If someone is
updating this not used file, then it is likely a bug because then it
shouldn't be listed in the unit file.

Lastly, systemd telling them to reload is not something that happens
automatically, so it's always up to the users discretion about when to
do it, and if they really want to update a particular listed file, but
never reload or get this message, they can always mask the service.
But I don't know of any scenarios where this would be wrong
information.

Cheers!

**Comment by abdulkareemnalband** [comment]
My two cents
How about extending `PATH UNIT` to do conditional reload of `SERVICE UNIT` instead of starting (activating)

**Comment by evverx** [comment]
@purpleidea , I'm not sure I understand
Will this

```
systemctl start some-service
touch /some/ConfigFile # for example: some people use :wq instead of :x in vim
```

produce annoying warnings?

**Comment by evverx** [comment]
Anyway, I think the last modified time doesn't really matter. We should check the file contents too (only?).
Similar to https://github.com/systemd/systemd/pull/3202#issuecomment-217482673

> But that's a lot of work, as that really requires comparing the old config with the new config to determine whether there was any change

Also, we should understand file formats.

```
A=1
B=2
```

is the same as

```
B=2
A=1
```

I guess this is not possible to implement in `pid1` (or doesn't make sense)

**Comment by purpleidea** [comment]
@evverx 

Indeed touching the config file and updating the lastModified time without actually changing any contents would produce the warning, but if you think about it, what business do people have poking around in the normally, and when they do, if they update the timestamp, then it's perfectly reasonable for us to notice it.

Keep in mind, we're talking about software specific config files here, not systemd unit files, so understanding the file formats is definitely out of scope for this, although storing a hash of the listed files, and comparing that would be a completely welcome extension to this proposal, but not something that I see as critical for the first iteration of this.

Thanks for your comments!

**Comment by johannbg** [comment]
This addition is something that should be put in it's own type unit file not as an extension of an existing one. 

There are alot of similar application stacks that require this or similar functionality thus adding this into existing type service units ( or type path units for that matter )  will make them messier and cannot be extended without them turning even messier. 

For example alot of people are using git "Push-to-Deploy" with post-receive hook that restarts type units ( amongs other things ) if new configuration file is being pushed or micro java service etc. so what he propose is precisely what they could take advantage of and drop that post-recieve git hook.

**Comment by johannbg** [comment]
Also note this can become racy if implemented incorrectly and should not be limited to a single file type extensions like .conf or whatever.

**Comment by nathwill** [comment]
the systemd primitive in play here seems like it ought to be the path unit. perhaps extending path units to be able to 'MarkForReload' or 'MarkDirty' the corresponding service (vs just activating as currently supported). that way checking the service status can still provide the service-reload-needed notification, and also provide a generic way (new dbus method?) to mark/query a service 'dirty' state?

**Comment by johannbg** [comment]
True type path units could be extended to cover this ( they are still simple enough ) I'm not sure why you would want to mark anything anyway why not just take action immediately? 

Unless of course that's your attempt to try to deal with the race condition as in mark it for reload or flag it dirty instead of immediately take action but if that's the case then arguably there is no point in adding this to systemd in the first place since the correct way at that point is to have whatever else is responsible for making those changes ( adding/editing/removing file(s) ) and have it tell systemd to start/restart/reload/stop type units when it thinks it's done ( you always have race condition one way or another if you try to use watch/openfile()readdir()inotify() and the more you hack the worse the race condition becomes  ) .

**Comment by nathwill** [comment]
> I'm not sure why you would want to mark anything anyway why not just take action immediately?

because vendors shouldn't be forcing reloads, but would still probably ship an enhanced path unit to help with tying config-file state to service 'dirty' state.

> if that's the case then arguably there is no point in adding this to systemd in the first place

my understanding from watching the mgmt talk was that the primary intent has to do with:
- the discovery (through a standard for vendor declaration) of the config files associated with a service. this can already _kind of_ be done through pkg inspection, but it's currently just an educated guess (files under /etc that the package providing the service unit file installs).
- knowing when a service is dirty _across cfg tool runs_ if the config file is changed in one run, but the tool crashes out before the service reload/restart is issued (typically at the end of the run in order to minimize the number required).

being able to mark a service as dirty immediately as a config file is changed solves the second problem; allowing vendors to declare their config files as triggers for services being marked dirty allows for a non-guesswork solution to the first problem.

so, it's a question of: watching files and storing service state as relates to 'reloaded since last cfg change'. 

for watching files, paths are the obviously applicable systemd primitive, and for storing dirty state, a property on the service unit seems like a simple, generic solution. a last-reloaded timestamp on the service unit would be helpful, but not as much as a 'Dirty' property to my way of thinking.

i may be missing something obvious, or misunderstanding the requestors intent, but insofar as i understand the issues at play, my suggestion is the most minimal and systemd-esque solution i could think of.

**Comment by johannbg** [comment]
I disagree 
Vendors or rather their application or application stack which is responsible for transferring/editing/deleting file(s) and ensuring their file(s) are in the correct state before service start/stop/restart/reload and even rollback if the service fails to start/restart/reload afterwards ( the config management application ), should be responsible for having systemd to take whatever action they deem fit should be done after doing so.  ( That's why git-based configuration deployments setups have been so successful, it takes care of all that headache for you and only starts/reload/restarts/stop the service and whatever post-hook action you have configured to run if it has run error free )

The thing is you don't need to flag type units dirty or for start/stop/restart/reload actions and you dont need to write additional tool to determine which state the type units is in after transferring/editing/deleting file(s) or if when compare when file(s) was last modified and the service startup/reload time if that's the case. 

James mentioned what config tool crashes and it will never know if the service needs reloading now I have to ask why is not the config tool flagging itself having failed in transaction and even in which transaction it failed ( if it's conducted in multiple steps ) so the action the tool took can be retried/rerun?  When does he actually expect the cms or the admin to notice to restart/reload after configuration changes if the cms is not notified immediately by itself through it's own failures ? 

Even if was successful in completion you always have to deal with something that can lead the service into failed state anyway ( corrupted configuration file or misconfiguration ) so despite systemd exposing the last modification of the file(s) of the file(s) the type path unit has been configured to monitor compare to when it was last started/restarted/reloaded/stopped. ( and you dont need to configure spesific state for that like "MarkForReload" systemd should just always suggest that if last modification time is newer than the active: state of the associated type service suggest reload ) you aren't gaining anything by it since upon the action of restarting/reloading the service, that service might fail anyway. 

So despite having something along the lines of ...
foo.path
[Unit]
Description=Foo CMS
Documentation= man:systemd.path man:cms(2)

[Path]
PathModified=/etc/foo.conf

[Install]
WantedBy=multi-user.target

foo.service
[Unit]
Description=Foo
Documentation=man:foo(1)

[Service]
ExecStart=/usr/bin/food 

[Install]
WantedBy=multi-user.target

and running systemctl status foo results in
foo.service - Foo
    Loaded: loaded (/usr/lib/systemd/system/gdm.service; enabled; vendor preset: enabled)
    Active: active (running) since Sun 2016-10-09 01:02:58 GMT; 21h ago
    Info: Associated file(s) have changed on disk. perhaps restart/reload is needed?
Main PID: 1234(foo)
....

I have a hard time seeing how this is actually supposed to solve anything for administrators in practice.

**Comment by poettering** [comment]
The mtime is the only thing we should consider. I mean, this is how these things are done on UNIX (see "make" for example), and it's what we already do to detect if systemd's own unit files are out-of-date.

Yes, if you touch a file without changing it you'll get the warning, but well: don't do it if you don't want that...

**Comment by purpleidea** [comment]
> discovery (through a standard for vendor declaration) of the config files associated with a service. this can already kind of be done through pkg inspection, but it's currently just an educated guess (files under /etc that the package providing the service unit file installs).

@nathwill exactly right, although as you mention, it's at best a guess because many packages don't ship all the files that should be listed. Best to engineer this properly and let them specify what they think matters to reload in the unit file.

> i may be missing something obvious, or misunderstanding the requestors intent, but insofar as i understand the issues at play

@nathwill Based on your comments in https://github.com/systemd/systemd/issues/4239#issuecomment-252509688 and my reading of them, I think you seem to understand the issue well. If you have any more questions feel free to ask online or offline.

> provide a generic way (new dbus method?) to mark/query a service 'dirty' state?

@nathwill It's an interesting idea, but I don't think it's something that will be useful. Let's consider the idea as a separate RFE if someone comes up with a use case. I personally can't think of a use case at this moment.

> James mentioned what config tool crashes and it will never know if the service needs reloading now I have to ask why is not the config tool flagging itself having failed in transaction and even in which transaction it failed ( if it's conducted in multiple steps ) so the action the tool took can be retried/rerun? When does he actually expect the cms or the admin to notice to restart/reload after configuration changes if the cms is not notified immediately by itself through it's own failures ? 

@johannbg Know the state of a service independent of config management or otherwise is a useful goal. In this case it will benefit config management. Keep in mind that in addition to tracking the state, this also lets us know the dependencies between services and their config files.

Thanks!

**Comment by johannbg** [comment]
@purpleidea you should be referring to your proposal as services and it's files since it does not seem to be limited to the daemon/services configuration files.

You can already express that in type path units as @nathwill pointed out ( and should be extended if some action is missing like take no action or just notify ) or ( since this will apparently never been gotten rid of ), you can express that as an EnvironmentFile=- entry in type service directly for soft association or ConditionPathExists= for hard dependency but you want to do more than just associating a set of files to a given service which would express that dependency right?

What do you feel like is missing from the type path units to cover your usecase?

**Comment by keszybz** [comment]
Just one option should be enough:

```
ConfigurationFiles=/etc/foo.conf /etc/apache/*.conf /etc/systemd/system/
```

I don't even think a separate section is needed, it can go in [Unit]. It seems similar to the Documentation= stanza.

The interpretation would be: if the argument is a directory, then consider all files underneath recursively. Otherwise, the argument is a glob pattern, consider all files that it matches. Single-file paths are a special kind of pattern, so need to distinguish them.

Seems simple enough and useful, so I'm +1 to the general idea.

I don't understand how the relative paths would work: in general every program which accepts multiple locations for config files uses a custom set of prefix directories, so I don't see how it those prefixes could be added automatically by systemd.

**Comment by johannbg** [comment]
@keszybz what's the difference between EnvironmentFile= and ConfigurationFiles= and why introduce another option instead of using that existing one?

**Comment by johannbg** [comment]
EnvironmentFile= should never have been introduced in the first place since the need for environment files became obsoleted with type units and if I would have continued the integration of systemd into Fedora I would have obsoleted all that stuff but they can gain an actual useful purpose being used for this.

**Comment by poettering** [comment]
So, my plan regarding all of this is to add a couple of new statements for declaring specific kinds of resources of a service, matching the RuntimeDirectory= setting we already have. Maybe something like this:

```
ConfigurationDirectory=
StateDirectory=
LogDirectory=
CacheDirectory=
```

Each of these settings would take a filename, that will be taken relative to the right prefix for the type of directory. Specifically, if you specify "foo" on `ConfigurationDirectory=` this would be translated to `/etc/foo` since `/etc` is the common prefix for configuration is /etc after all. Similar: `StateDirectory=foo` would become `/var/lib/foo`, `LogDirectory=foo` would become `/var/log/foo` and `CacheDirectory=foo` would become `/var/cache/foo`.

That's what would happen for system services. For user services it would instead become ~/.config/foo (for the first three listed above), ~/.cache/foo (for CacheDirectory).

With all that in place we could then impement the hint about "configuration out of date" (by simply traversing the config directory until we find at least one file that is newer than the last reload time). We could also use it to implement a new "systemctl purge" command, that can flush out configuration, state, logs or cache of a specific service. Finally, we'd implicitly add these dirs as writable to the sandboxing (i.e. imply WriteDirectory= for them), and add some special magic to make sure they play nicely with DynamicUser= (i.e. that we try to reuse the owner uid of the `/var/lib/<foo>` directory, and if it is busy chown the subdir).

**Comment by johannbg** [comment]
Are you going to add all that to type service unit files or are you going to introduce a new type environment unit where all this can be declared?

**Comment by poettering** [comment]
> Are you going to add all that to type service unit files or are you going to introduce a new type environment unit where all this can be declared?

The `XyzDirectory=` settings I proposed above would be declared in service files, within the `[Service]` section, exactly like `RuntimeDirectory=` is already declared there.

**Comment by johannbg** [comment]
@poettering why not introduce all of this in a new type unit since you most likely will need to introduce this and more for portals no? 
What are the pros/cons you see in having this in the type service units as opposed to a separated environment file?

**Comment by pspacek** [comment]
Side-note: Tracking all configuration files which are significant for particular program/service gets interesting. E.g. programs using NSS library might need to list /etc/nsswitch.conf or /etc/resolv.conf - similarly to any other configuration file used by a library called by the main program...

**Comment by johannbg** [comment]
Right this will get inherently messy quite fast,increase system resource usage and complicate administrations significantly.
Vendor and upstream will have to take care declaring the correct configuration directory entries for /usr/lib/$foo files and directories for their component ( including their directory .d counterparts ) as well as you point out also for the other component that it might rely on or use other components configuration files ( which will introduce dependencies and correct ordering ).

Administrators will have to create vendor overwriting configuration snippets and take care of declaring configurations for /etc/foo, /usr/local/foo, /srv/foo or /opt/foo ( including their directory .d counterparts ). 

Systemd will have to come up with some kind of policy with the correct precedence out of the box with an opt overwrite toggle for upstream and or vendors as well as administrators to declare the correct precedence ordering on top of ensuring something like tmpfile does not trigger when it generates /etc/foo configuration files(s) and directorie(s), 

And since @poettering  plans to implement this in type service units it will make type service units even a more pigs mess than it already has evolved into o_O. 
All of which done to satisfy a single file type of configuration type for an Red Hat product which seems to rely on packagekit out of all things present and may or may not take off in the future.

**Comment by purpleidea** [comment]
On Thu, Nov 3, 2016 at 6:18 AM, Jóhann B. Guðmundsson <
notifications@github.com> wrote:

> All of which done to satisfy a single file type of configuration type for
> an Red Hat product which may or may not take off in the future.

If you're referring to mgmt, it is _not_ a "Red Hat product", it is a
_community_ project, completely free software, no open core, and developed
in the open. Please join us! In addition, the proposal benefits other
tools, both legacy and future looking, as well as the sysadmin using
systemctl without any add ons.

I'm happy to discuss with your privately if you'd prefer to take this off
list. Ping me on IRC, I'm in #mgmtconfig on Freenode.

Cheers

**Comment by poettering** [comment]
I figure we can close this now that we have ConfigurationDirectory=

**Comment by purpleidea** [comment]
@poettering Awesome, thanks for the heads up!

In looking at the patch, AFAICT, the "features" part of showing mtime is greater in systemctl, and so on isn't in yet. Is there a different tracker for that stuff?

Does this allow specifying multiple ConfigurationDirectory path's, potentially even including single files? (Like /etc/drbd.conf for example) ?

My understanding is that I can query that unit file properties via dbus, although I haven't tried this patch out yet.

I'm happy to write a blog post encouraging folks to add these properties to the upstream unit files if you'd like.

Thanks!

**Comment by poettering** [comment]
> In looking at the patch, AFAICT, the "features" part of showing mtime is greater in systemctl, and so on isn't in yet. Is there a different tracker for that stuff?

Not sure I parse this sentence properly, but showing "configuration changed" info based on mtime isn't implemented yet. Please file an issue requesting that. Or even better a PR adding that. ;-)



> Does this allow specifying multiple ConfigurationDirectory path's, potentially even including single files? (Like /etc/drbd.conf for example) ?

It only supports specifying directories, as many as you like. It does not support specifying files. And I don't think we should change that. Daemons really should use their own neatly separated dirs for their config stuff, much like they should use their own dirs below /var/run, /var/lib and so on.



> My understanding is that I can query that unit file properties via dbus, although I haven't tried this patch out yet.

Yes, that's correct.
